### PR TITLE
Discord Proposal Demo for easier Discord signup 

### PR DESCRIPTION
**Pull request by elenabaurkot**
### Description
Created a page in Platform for fellows to sign up for Discord more easily. This is just for the demo on Friday and it will be cleaned up more later.

### Motivation and Context
Proposal:
- Stop giving people Discord invite links in the welcome email
- Add a page in Canvas/Platform to instruct people to create a Discord account at https://discord.com/register , or sign in to their existing account if they have one. (If this needs to happen before people get into canvas, we can make a Platform page that's accessible even without a canvas account. But that's less ideal.) (we can also prefill an email on this page if we want, like https://discord.com/register?email=test@example.com - BUT that will cause more pain if it asks them for a phone number and they enter one they've used before with a different Discord account.)
- On the same page mentioned above, add a button they can click to authorize access to their Discord account (https://discord.com/developers/docs/topics/oauth2#oauth2), like we do with LinkedIn in the H2C project.
- From there, we can read their email address and whether or not they've verified that email, and give them instructions on how to add or verify an email if we detect they haven't (https://discord.com/developers/docs/resources/user)
- Once we're sure they have everything set up correctly, we add them to the server programmatically (https://discord.com/developers/docs/resources/guild#add-guild-member), and give them a direct (non-invite) link to the server. (Staff never have to generate new invites, because participants don't use invites at all.)

### Screenshots (if appropriate):
Screen Shot 2021-10-14 at 3 59 49 PM

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: app/views/discord_signup/launch.html.erb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+<% if @discord_user.nil? %>
+  <p>Complete step one, then come back to complete the following steps.</p>
+
+  <h2>Step 1: Sign up/Register</h2>
+  <p>If you have a Discord account sign in here: <a href="https://discord.com/login">https://discord.com/login</a></p>
+
+  <p>If you don't have a Discord account register here: <a href="https://discord.com/register">https://discord.com/register</a></p>
+
+  <h2>Step 2: Authorize Your Discord Account</h2>
+  <p><a href="https://discord.com/api/oauth2/authorize?client_id=847515831285973033&redirect_uri=https%3A%2F%2Felenaplatform.ngrok.io%2Fdiscord_signup%2Foauth&response_type=code&scope=identify%20email%20guilds%20guilds.join">Authorize Braven to access your Discord account</a></p>
```
Comment (rshipp) [inline app/views/discord_signup/launch.html.erb:13]:
```suggestion
  <p><a href="https://discord.com/api/oauth2/authorize?client_id=<%= Rails.application.secrets.discord_client_id %>&redirect_uri=<%= ERB::Util.url_encode(Rails.application.secrets.discord_redirect_uri) %>&response_type=code&scope=identify%20email%20guilds%20guilds.join">Authorize Braven to access your Discord account</a></p>
```

File: app/assets/stylesheets/discord_signup.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the DiscordSignup controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (sadleb) [inline app/assets/stylesheets/discord_signup.scss:3]: You need to add the following to `app/assets/stylesheets/layouts/lti_canvas.scss` in order for it to work in prod:
```
//*= require discord_signup
```

File: app/controllers/discord_signup_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+  def launch
+    authorize :discord_signup
+
+    if current_user.discord_token
+      response = Discordrb::API::User.profile("Bearer #{current_user.discord_token}")
+      @discord_user = JSON.parse(response.body)
+
+      if @discord_user["email"] && @discord_user["verified"]
+        Discordrb::API::Server.add_member("Bot #{Rails.application.secrets.bot_token}", "722098701878689933", @discord_user["id"], current_user.discord_token)
```
[the remark this answers is not in the document]
Comment (rshipp) [reply app/controllers/discord_signup_controller.rb:14]: server id, and no it's just for our test server. would eventually be replaced by dynamic IDs depending on which region the Fellow is in

File: app/views/discord_signup/launch.html.erb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    <p>If you don't see the email, go to your <a href="https://discord.com/channels/@me">Discord home page </a></p>
+
+    <p>Click the gear in the lower left corner to go to your user settings.</p>
+    <img src="https://res.cloudinary.com/dsxuuory9/image/upload/v1634239612/discorddemo/Screen_Shot_2021-10-14_at_3.26.45_PM_sgdr3l.png" alt="" style="max-width: 100%; margin: 0 auto 35px auto; display: block;">
+
+    <p>At the top of the user settings , click the button that says "Resend Verification Email"</p>
+    <img src="https://res.cloudinary.com/dsxuuory9/image/upload/v1634239519/discorddemo/Screen_Shot_2021-10-14_at_2.21.15_PM_irmrvy.png" alt="" style="max-width: 100%; margin: 0 auto 35px auto; display: block;">
+  <% end %>
+
+  <p style="padding-bottom: 50px;">Once you're all set up, continue to the <a href="https://discord.com/channels/722098701878689933/846463735961878598">Braven Discord Server</a></p>
```
Comment (sadleb) [inline app/views/discord_signup/launch.html.erb:50]: Will this inline style break in prod?

File: config/environments/development.rb [hunk 5]
```diff
@@ -14,7 +14,7 @@
   # Show full error reports.
   config.consider_all_requests_local = true
   # Set this to false to show custom error views in development.
-  #config.consider_all_requests_local = false
+  # config.consider_all_requests_local = false
```
Comment (sadleb) [inline config/environments/development.rb:17]: Doesn't matter too much, but maybe revert this change so this PR / commit doesn't show in the git blame / history for this file?

File: app/views/discord_signup/launch.html.erb (hunk 4 above)
Comment (rshipp) [reply app/views/discord_signup/launch.html.erb:50]: ugh, it might, I forgot about the CSP.

File: app/controllers/discord_signup_controller.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def oauth
+    authorize :discord_signup
+
+    response = Discordrb::API.request(:oauth2_token, nil, :post, 'https://discord.com/api/v8/oauth2/token', {client_id: Rails.application.secrets.discord_client_id, client_secret: Rails.application.secrets.discord_client_secret, grant_type: 'authorization_code', code: params[:code], redirect_uri: Rails.application.secrets.discord_redirect_uri}, content_type: 'application/x-www-form-urlencoded')
+
+    current_user.update!(discord_token: JSON.parse(response.body)["access_token"])
+
+    redirect_to launch_discord_signup_index_url
```
[the remark this answers is not in the document]
Comment (rshipp) [reply app/controllers/discord_signup_controller.rb:26]: As for why it has an `index` in it, i have no idea, that's just what Rails did. 

Screen Shot 2021-10-14 at 4 28 48 PM

File: app/assets/stylesheets/discord_signup.scss (hunk 2 above)
Comment (elenabaurkot) [reply app/assets/stylesheets/discord_signup.scss:3]: ah okay, got it, thank you @sadleb

File: config/secrets.yml [hunk 7]
```diff
@@ -37,6 +37,10 @@ development:
   support_ticket_url: <%= ENV.fetch('SUPPORT_TICKET_URL') { 'https://form.123formbuilder.com/5811830' } %>
   zoom_api_key: <%= ENV['ZOOM_API_KEY'] %>
   zoom_api_secret: <%= ENV['ZOOM_API_SECRET'] %>
+  discord_client_id: <%= ENV['DISCORD_CLIENT_ID'] %>
+  discord_client_secret: <%= ENV['DISCORD_CLIENT_SECRET'] %>
+  discord_redirect_uri: <%= ENV['DISCORD_REDIRECT_URI'] %>
+  bot_token: <%= ENV['BOT_TOKEN'] %>
```
[the remark this answers is not in the document]
Comment (sadleb) [reply config/secrets.yml:43]: kk. i suppose the renaming will be the forcing mechanism to remember to do this

File: app/controllers/discord_signups_controller.rb [hunk 8]
```diff
@@ -1,12 +1,12 @@
...
-class DiscordSignupController < ApplicationController
+class DiscordSignupsController < ApplicationController
   layout 'lti_canvas'
 
   def launch
-    authorize :discord_signup
+    authorize :discord_signups
 
-    if current_user.discord_token
+    if !current_user.discord_token
```
Comment (sadleb) [inline app/controllers/discord_signups_controller.rb:9]: Was this on purpose @rshipp ? Seems wrong?

Comment (rshipp) [reply app/controllers/discord_signups_controller.rb:9]: Pushed directly to highlander d574aabe43530512ad4c3d4e95709810533e745f, thanks for catching this
